File: IMCA398/node/imgDL.mjs
// import fetch from 'node-fetch';
// const response = await fetch('https://github.com/');
// console.log(response);
// const body = await response.text();
// console.log(body);

// import fs from "fs";
// import fetch from 'node-fetch';
// const downloadImage = async (url, path) => {
//     const response = await fetch(url);
//     const blob = await response.blob();
//     const arrayBuffer = await blob.arrayBuffer();
//     const buffer = Buffer.from(arrayBuffer);
//     await fs.writeFile(path, buffer);
// }
// await downloadImage("https://sabe.io/images/saturn.png", "imgDL/saturn.png")
// // TypeError[ERR_INVALID_CALLBACK]: Callback must be a function. Received undefined

import fs from "fs";
import fetch from 'node-fetch';

const downloadImage = async (url, path) => {
    const response = await fetch(url);
    // console.log(`response:` + response)
    // console.log(response)
    const blob = await response.blob();
    // console.log(`blob:` + blob)
    // console.log(blob)
    const arrayBuffer = await blob.arrayBuffer();
    // console.log(`arrayBuffer:` + arrayBuffer)
    // console.log(arrayBuffer)
    const buffer = Buffer.from(arrayBuffer);
    console.log(buffer)
    await fs.writeFile(path, buffer, function (err, result) {
        if (err) console.log('error', err);
    });
}

// await downloadImage("https://sabe.io/images/saturn.png", "imgDL/saturn.png")
// await downloadImage("https://replicate.delivery/pbxt/Ib0vihEn6spRBdzLgoEez8hHo3jFMdCZawEeEaueJb4ekGLCB/out-0.png", "imgDL/1.jpg")
// await downloadImage("https://replicate.delivery/pbxt/Ib0vihEn6spRBdzLgoEez8hHo3jFMdCZawEeEaueJb4ekGLCB/out-0.png", "imgDL/test.png")
await downloadImage("https://replicate.delivery/pbxt/TUkoT15HUO4jAtTt1hdId1uRDg3ovAo6uw4KupVaybHom0IE/out-0.png", "imgDL/1.jpg")

// fs.createReadStream(path, options)

File: IMCA398/python/imgDL.py

# from urllib.request import urlretrieve
# url = "https://replicate.delivery/pbxt/nJ9UM4x9zCb3P98ljgd8qJtcBA7ke9UJeOY4aIC0Ac3BlYgQA/out-0.png"
# urlretrieve(url, "/tmp/out.png")
# background = Image.open("/tmp/out.png")
# # ERR: SSL: CERTIFICATE_VERIFY_FAILED

# import urllib.request, base64
# url = "https://replicate.delivery/pbxt/nJ9UM4x9zCb3P98ljgd8qJtcBA7ke9UJeOY4aIC0Ac3BlYgQA/out-0.png"
# request = urllib.request(url)
# request.add_header("Authorization", "Token b590de6e35d35db67acdeee0633d98708e705124")   
# result = urllib.urlopen(request)
# # ERR: TypeError: 'module' object is not callable

# from urllib.request import urlretrieve
# url = "https://replicate.delivery/pbxt/nJ9UM4x9zCb3P98ljgd8qJtcBA7ke9UJeOY4aIC0Ac3BlYgQA/out-0.png"
# retrieve = urlretrieve(url, "/tmp/out.png")
# retrieve.add_header("Authorization", "Token b590de6e35d35db67acdeee0633d98708e705124")  
# background = Image.open("/tmp/out.png")
# # ERR: SSL: CERTIFICATE_VERIFY_FAILED

# import urllib.request
# url = "https://replicate.delivery/pbxt/nJ9UM4x9zCb3P98ljgd8qJtcBA7ke9UJeOY4aIC0Ac3BlYgQA/out-0.png"
# request = urllib.request.urlopen(url)
# request.add_header("Authorization", "Token b590de6e35d35db67acdeee0633d98708e705124")  
# # ERR: SSL: CERTIFICATE_VERIFY_FAILED
 
# import requests
# response = requests.get("https://replicate.delivery/pbxt/nJ9UM4x9zCb3P98ljgd8qJtcBA7ke9UJeOY4aIC0Ac3BlYgQA/out-0.png")
# # print(response) # <Response [200]>
# # print(response.url) # actual url of the image
# # print(response.content) # bytes
# # print(response.text) # ???
# print(response.json) # ???

# import urllib3
# url = "https://replicate.delivery/pbxt/nJ9UM4x9zCb3P98ljgd8qJtcBA7ke9UJeOY4aIC0Ac3BlYgQA/out-0.png"
# req = urllib3.request(url)
# req.add_header("Authorization", "Token b590de6e35d35db67acdeee0633d98708e705124")   
# result = urllib3.urlopen(req)
# # ERR: TypeError: 'module' object is not callable

# from https://stackoverflow.com/questions/63977849/recommended-way-to-download-images-in-python-requests
import requests
i = 2
url = "https://replicate.delivery/pbxt/Xci4OdhG4QLQJNFhfzJeQ73orBjfaEenSXs4G7NyvpxqU3BCB/out-0.png"
with open('imgDL/file{}.png'.format(i), 'wb') as f:
    f.write(requests.get(url).content)
# WORKS???

